modules/nlu/src/backend/cloud/oauth.ts

```diff
@@ -0,0 +1,71 @@
+import axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from 'axios'
+import { cache } from './cache'
+import qs from 'querystring'
+
+interface OauthTokenClientProps {
+  oauthUrl: string
+  clientId: string
+  clientSecret: string
+}
+
+interface OauthResponse {
+  access_token: string
+  expires_in: number
+  scope: string
+  token_type: string
+}
+
+const createOauthTokenClient = (
+  axios: AxiosInstance,
+  { oauthUrl, clientId, clientSecret }: OauthTokenClientProps
+) => () =>
+  axios
+    .post(
+      oauthUrl,
+      qs.stringify({ client_id: clientId, client_secret: clientSecret, grant_type: 'client_credentials' })
+    )
+    .then(res => res.data as OauthResponse)
+
+const requestInterceptor = (authenticate: () => Promise<string>) => async (config: AxiosRequestConfig) => {
+  const token = await authenticate()
+  config.headers.Authorization = `Bearer ${token}`
+  return config
+}
+
+type ErrorRetrier = AxiosError & { config: { _retry: boolean } }
+
+const errorInterceptor = (instance: AxiosInstance, authenticate: () => Promise<string>) => async (
+  error: ErrorRetrier
+) => {
+  if (error.response?.status === 401 && !error.config._retry) {
+    error.config._retry = true
+    const token = await authenticate()
+    const config = error.config
+    config.headers.Authorization = `Bearer ${token}`
+    return instance.request(config)
+  }
+
+  return Promise.reject(error)
+}
+
+export type OauthClientProps = OauthTokenClientProps & {
+  endpoint: string
+}
+
+export const createOauthClient = ({ clientId, clientSecret, endpoint, oauthUrl }: OauthClientProps) => {
+  const oauthTokenClient = createOauthTokenClient(axios.create(), {
+    oauthUrl,
+    clientId,
+    clientSecret
+  })
+
+  const tokenCache = cache(oauthTokenClient, {
+    getExpiry: res => res.expires_in * 1000,
```

Comment:
\*\*Personal pref\*\*:

put the unit of measure (ms or secs) in the function name:

```ts
const tokenCache = cache(oauthTokenClient, {
    getExpiryInMs: res => res.expires_in,
    ...
```